core/blockchain/impl/key_value_block_storage.cpp

```diff
@@ -74,24 +78,76 @@ namespace kagome::blockchain {
 
   outcome::result<primitives::BlockBody> KeyValueBlockStorage::getBlockBody(
       const primitives::BlockId &id) const {
-    OUTCOME_TRY(encoded_body, getWithPrefix(*storage_, Prefix::BODY, id));
-    OUTCOME_TRY(body, scale::decode<primitives::BlockBody>(encoded_body));
-    return std::move(body);
+    OUTCOME_TRY(block_data, getBlockData(id));
+    if (block_data.body) {
+      return block_data.body.value();
+    }
+    return Error::BODY_DOES_NOT_EXIST;
+  }
+
+  outcome::result<primitives::BlockData> KeyValueBlockStorage::getBlockData(
+      const primitives::BlockId &id) const {
+    OUTCOME_TRY(encoded_block_data,
+                getWithPrefix(*storage_, Prefix::BLOCK_DATA, id));
+    OUTCOME_TRY(block_data,
+                scale::decode<primitives::BlockData>(encoded_block_data));
+    return std::move(block_data);
   }
 
   outcome::result<primitives::Justification>
   KeyValueBlockStorage::getJustification(
       const primitives::BlockId &block) const {
-    OUTCOME_TRY(encoded_just,
-                getWithPrefix(*storage_, Prefix::JUSTIFICATION, block));
-    OUTCOME_TRY(just, scale::decode<primitives::Justification>(encoded_just));
-    return std::move(just);
+    OUTCOME_TRY(block_data, getBlockData(block));
+    if (block_data.justification) {
+      return block_data.justification.value();
+    }
+    return Error::JUSTIFICATION_DOES_NOT_EXIST;
+  }
+
+  outcome::result<primitives::BlockHash> KeyValueBlockStorage::putBlockHeader(
+      const primitives::BlockHeader &header) {
+    OUTCOME_TRY(encoded_header, scale::encode(header));
+    auto block_hash = hasher_->blake2b_256(encoded_header);
+    OUTCOME_TRY(putWithPrefix(*storage_,
+                              Prefix::HEADER,
+                              header.number,
+                              block_hash,
+                              Buffer{encoded_header}));
+    return block_hash;
+  }
+
+  outcome::result<void> KeyValueBlockStorage::putBlockData(
+      primitives::BlockNumber block_number,
+      const primitives::BlockData &block_data) {
+    primitives::BlockData to_insert;
+
+    // if block data does not exist, put a new one. Otherwise get the old one
+    // and merge with the new one
+    auto existing_block_data_res = getBlockData(block_data.hash);
+    if (not existing_block_data_res) {
+      to_insert = block_data;
+    } else {
+      to_insert = existing_block_data_res.value();
+
+      // add all the fields from the new block_data
+      to_insert.header = block_data.header;
+      to_insert.body = block_data.body;
+      to_insert.justification = block_data.justification;
+      to_insert.message_queue = block_data.message_queue;
+      to_insert.receipt = block_data.receipt;
+    }
```

Comment:
Better to use move-assignment

Reply:
on what? block data is a const reference

Reply:
ouch.